### PR TITLE
[wgsl-in] Fail on repeated attributes

### LINKED ISSUE #2425: [wgsl-in] attributes appearing more than once are not rejected
This shader compiles:
```wgsl
struct FragmentIn {
    @location(0) @location(1) color: vec4<f32>,
}

@fragment
fn main(in: FragmentIn) -> @location(0) @location(1) vec4<f32> {
    return vec4<f32>(in.color.rgb, in.color.a);
}
```

even though the `location` attribute appears more than once on the struct member and the return type of the entry point (in this case, the last specified location appears to be taken into account). The WGSL spec says in [11. Attributes](https://www.w3.org/TR/WGSL/#attributes):

> Unless explicitly permitted below, an attribute [must not](https://www.w3.org/TR/WGSL/#shader-creation-error) be specified more than once per object or type.

I'm not sure whether this means a diagnostic is required, or even that the compilation should fail.

### PR DESCRIPTION
**Pull request by fornwall**
Fixes #2425.

### CODE REVIEW

**Review by ErichDonGubler — APPROVED** [review]
LGTM, minus some nits!

File: src/front/wgsl/error.rs [hunk 1]
```diff
@@ -136,6 +136,7 @@ pub enum InvalidAssignmentType {
 }
 
 #[derive(Clone, Debug)]
+#[cfg_attr(test, derive(PartialEq))]
```
[the remark this answers is not in the document]
Comment (fornwall) [reply src/front/wgsl/error.rs]: Thanks! Fixed in fb1911108de1c60ebc1d88557a41ef9ea4653de7

File: src/front/wgsl/tests.rs [hunk 2]
```diff
@@ -481,3 +481,30 @@ fn parse_alias() {
...
+
+#[test]
+fn parse_repeated_attributes() {
+    use crate::{
+        front::wgsl::{error::Error, Frontend},
+        Span,
+    };
+
+    let template = "@vertex fn vs() -> __REPLACE__ vec4<f32> { return vec4<f32>(0.0); }";
+    for attribute in [
```
[the remark this answers is not in the document]
Comment (fornwall) [reply src/front/wgsl/tests.rs:493]: Makes sense? Do you think we should add something like [enum_derive](https://docs.rs/enum_derive/latest/enum_derive/) to iterate over values in tests, or do you have another preferred variant?

**Comment by teoxoy** [comment]
There are more attributes that this validation should cover. I think all attributes can't be repeated (the spec mentions to look for exceptions to this rule for each attribute but at a first glance I can't see any such exception).

**Comment by fornwall** [comment]
> There are more attributes that this validation should cover. I think all attributes can't be repeated (the spec mentions to look for exceptions to this rule for each attribute but at a first glance I can't see any such exception).

Thanks @teoxoy ! I've added additional attributes in 1a7c3abb0dfbea055be15865fbe72fc7e1bab0dc

`workgroup_size` is missing. I'm planning to do that in a separate PR (to propose changing `EntryPoint.workgroup_size` from `[u32; 3]` to `Option<[u32; 3]>`, which will also fix #2271). Since that is a bigger change touching all backends and frontends, is that ok?

File: src/front/wgsl/tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fornwall) [reply src/front/wgsl/tests.rs]: OTOH that should probably come as a follow up PR as you say, to decrease scope.

**Comment by fornwall** [comment]
Rebased to fix merge conflicts in `src/front/wgsl/tests.rs`.

**Comment by teoxoy** [comment]
Sounds good!

Given that all the attributes are an `Option<Something>`, I feel like a more robust solution would be to introduce a newtype that once you set its inner `Option` via a `set` method, you'll keep getting the repeated attributes error on subsequent `set` calls. Thoughts?

**Comment by fornwall** [comment]
> Given that all the attributes are an `Option<Something>`, I feel like a more robust solution would be to introduce a newtype that once you set its inner `Option` via a `set` method, you'll keep getting the repeated attributes error on subsequent `set` calls. Thoughts?

Sounds like a good idea, I'll try it out!

**Comment by fornwall** [comment]
> Given that all the attributes are an Option<Something>, I feel like a more robust solution would be to introduce a newtype that once you set its inner Option via a set method, you'll keep getting the repeated attributes error on subsequent set calls. Thoughts?

@teoxoy Like this: f09f9f70f3fc834cafc41ea9bc1aa00eed13bdf5?

**Review by teoxoy — APPROVED** [review]
LGTM, thanks for making the change!

**Comment by teoxoy** [comment]
@ErichDonGubler want to give this another look? 👀

**Comment by ErichDonGubler** [comment]
Feeling silly that my unfamiliarity with WGSL led me to miss other attributes, but otherwise, this LGTM. Could we file an issue for getting `enum`-exhaustive coverage for these attributes for testing before we merge? I think that that's the only preference I want to call out.

**Comment by fornwall** [comment]
> Could we file an issue for getting enum-exhaustive coverage for these attributes for testing before we merge? I think that that's the only preference I want to call out.

Thanks @ErichDonGubler! Have added https://github.com/gfx-rs/naga/issues/2433 as a follow up issue for that, and https://github.com/gfx-rs/naga/issues/2434 for the `@workgroup_size` remaining attribute (which is a bit bigger, since making workgroup_size an `Option` will touch more parts).

**Comment by teoxoy** [comment]
Thanks! Will merge this then.

**Pull request merged**